api/src/pcapi/core/educational/api/shared.py

```diff
@@ -0,0 +1,79 @@
+import datetime
+
+from pcapi.core.educational import exceptions as educational_exceptions
+from pcapi.core.educational import models as educational_models
+from pcapi.core.educational import repository as educational_repository
+from pcapi.core.educational import utils as educational_utils
+from pcapi.models import feature
+
+
+def update_collective_stock_booking(
+    stock: educational_models.CollectiveStock,
+    current_booking: educational_models.CollectiveBooking | None,
+    beginning_has_changed: bool,
+) -> None:
+    """When a collective stock is updated, we also update some fields of its related booking"""
+
+    # if the booking limit date is set in the future and a PENDING booking was automatically cancelled, set it back to PENDING
+    last_booking = stock.lastBooking
+    booking_limit_value = stock.bookingLimitDatetime
+    expired_booking_to_update = None
+    if _should_update_collective_booking_pending(last_booking, booking_limit_value):
+        assert last_booking is not None  # checked by should_update_collective_booking_pending
+        expired_booking_to_update = last_booking
+        _update_collective_booking_pending(last_booking)
+
+    # the booking to update should either be the expired booking that was set back to PENDING
+    # or the non-cancelled booking (there should not be one of each at the same time)
+    if expired_booking_to_update and current_booking:
+        raise educational_exceptions.MultipleCollectiveBookingFound()
+
+    booking_to_update = expired_booking_to_update or current_booking
+    if booking_to_update:
+        booking_to_update.confirmationLimitDate = booking_limit_value
+
+        if beginning_has_changed:
+            _update_collective_booking_cancellation_limit_date(booking_to_update, stock.beginningDatetime)
+            _update_collective_booking_educational_year_id(booking_to_update, stock.beginningDatetime)
+
+
+def _update_collective_booking_educational_year_id(
+    booking: educational_models.CollectiveBooking,
+    new_beginning_datetime: datetime.datetime,
+) -> None:
+    educational_year = educational_repository.find_educational_year_by_date(new_beginning_datetime)
+    if educational_year is None:
+        raise educational_exceptions.EducationalYearNotFound()
+
+    booking.educationalYear = educational_year
+
+
+def _update_collective_booking_cancellation_limit_date(
+    booking: educational_models.CollectiveBooking, new_beginning_datetime: datetime.datetime
+) -> None:
+    # if the input date has a timezone (resp. does not have one), we need to compare it with an aware datetime (resp. a naive datetime)
+    now = (
+        datetime.datetime.utcnow()
+        if new_beginning_datetime.tzinfo is None
+        else datetime.datetime.now(datetime.timezone.utc)  # pylint: disable=datetime-now
```

Comment:
Mais du coup il ne faudrait pas comparer avec `new_beginning_datetime.tzinfo` l'input date est tz aware ?

Reply:
Via la route Pro on enlève la TZ
Via la route publique on a la TZ sur la date

Ici on permet de comparer avec now avec ou sans TZ suivant si `new_beginning_datetime` a une TZ ou non

Pour moi si 2 dates ont des TZ différentes on est capable de les comparer, ce qu'il ne faut pas c'est une date avec et une date sans

Reply:
Bah `now` en UTC, c'est pas du tout la même chose que `now` en Guadeloupe, et in fine ça influe sur le calcul de la date limite d'annulation:

```python
In [1]: import pytz

In [2]: import datetime

In [3]: datetime.datetime.now(pytz.timezone("UTC"))
Out[3]: datetime.datetime(2024, 12, 6, 10, 14, 53, 600092, tzinfo=<UTC>)

In [4]: datetime.datetime.now(pytz.timezone("America/Guadeloupe"))
Out[4]: datetime.datetime(2024, 12, 6, 6, 14, 58, 701813, tzinfo=<DstTzInfo 'America/Guadeloupe' AST-1 day, 20:00:00 STD>)
```

Je vois que c'est déjà comme ça sur master et que c'est juste du code déplacé, laissons comme ça du coup, c'est pas vraiment du scope de ta PR

Reply:
Yes c'est déjà en place

Mais pour moi tu vas recevoir ton new_beginning_datetime avec une TZ, disons Guadeloupe

Si tu compares ça à now en UTC, la lib datetime va (il me semble) ramener la new_beginning_datetime en UTC (en changeant l'heure en conséquence) et comparer, ce qui revient à comparer le new_beginning_datetime de départ en TZ Guadeloupe avec now en TZ Guadeloupe